### PR TITLE
Feature/audit-system

### PR DESCRIPTION
**Pull request by MohamedEmad7299**
This PR introduces a complete audit tracking system for tasks and projects in PlanMate, allowing users to retrieve a detailed change history by project ID or task ID. Each audit entry includes:

 - Who made the change

 - What was changed (e.g., status updates)

 - When the change occurred

Example:
User abc changed task XYZ-001 from InProgress to InDevReview at 2025/05/24 8:00 PM

Key Features 

 - Fully integrated audit trail logic

 - Track changes for both tasks and projects

 - Designed to support future enhancements (e.g., filtering, pagination, etc.)

Test Coverage

✅ 100% test coverage across all new logic
✅ Includes extensive unit tests for edge cases and expected behavior
✅ All tests passed successfully

### CODE REVIEW

**Review by eslam-99 — CHANGES_REQUESTED** [review]
why all these files, please fix merge on your branch to avoid modifying all these files

**Comment by MohamedEmad7299** [comment]
> why all these files, please fix merge on your branch to avoid modifying all these files

To make test cover 100% of the branch as a abdallah said

**Comment by eslam-99** [comment]
> > why all these files, please fix merge on your branch to avoid modifying all these files
> 
> To make test cover 100% of the branch as a abdallah said

this already modified in separate branch and merged to dev, no need for any other branches to do so, just when merging accepts your files of the task instead of empty ones

File: src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        )
+
+        // when
+        useCase.addAudit(
+            audit.id, audit.createdBy, audit.entityId, audit.entityType,
+            audit.oldState, audit.newState, audit.date
+        )
+
+        // then
+        verify(exactly = 1) { auditRepository.addAuditLog(audit) }
```
Comment (AmrAshraf-git) [inline src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt]: Use match or `verify { addAuditLog(match { ... }) }` to compare object fields rather than identity
reason: -
You're verifying that `auditRepository.addAuditLog(audit)` was called with exactly the same instance you created in test (`val audit = createAudit(...)`).
But in the actual AddAuditUseCase, a new Audit object is constructed inside the method, so the object passed to `addAuditLog()` will not be the same instance.

**Review by eslam-99 — DISMISSED** [review]
Why deleting models?
already added to be excluded from testing, so no worries

File: src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt (hunk 1 above)
Comment (MohamedEmad7299) [reply src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt]: Thank U for your useful insight 
SOLVED!

File: src/main/kotlin/logic/audit/AddAuditUseCase.kt [hunk 2]
```diff
@@ -1,2 +1,44 @@
...
+import squad.abudhabi.logic.model.Audit
+import squad.abudhabi.logic.model.EntityType
+import squad.abudhabi.logic.repository.AuditRepository
+import java.time.LocalDate
+
+class AddAuditUseCase(
+    private val auditRepository: AuditRepository
+) {
+    fun addAudit(
+        id: String,
```
Comment (nourelhodaahmed) [inline src/main/kotlin/logic/audit/AddAuditUseCase.kt]: why not make the id have value by default in data class
val id: String= UUID.randomUUID().toString(),
and not take it here as param

File: src/main/kotlin/logic/audit/AddAuditUseCase.kt [hunk 3]
```diff
@@ -1,2 +1,44 @@
...
+        )
+
+        if (auditLog.isValid() && !isSameStates(newState = newState, oldState = oldState))
+        auditRepository.addAuditLog(auditLog)
+        else throw InvalidAudit()
+    }
+
+    private fun Audit.isValid(): Boolean {
+        return newState.isNotEmpty() &&
+               id.isNotEmpty() &&
```
Comment (nourelhodaahmed) [inline src/main/kotlin/logic/audit/AddAuditUseCase.kt]: when you add id in data class, remove this line as it will never be empty

File: src/main/kotlin/logic/audit/GetAuditUseCase.kt [hunk 4]
```diff
@@ -1,2 +1,16 @@
...
+import squad.abudhabi.logic.repository.AuditRepository
+
+
+class GetAuditUseCase(
+    private val auditRepository: AuditRepository
+){
+
+    fun getAuditHistory(entityId: String): List<Audit> {
+
+        return if (entityId.isNotEmpty()) auditRepository.getAuditByEntityId(entityId)
```
Comment (nourelhodaahmed) [inline src/main/kotlin/logic/audit/GetAuditUseCase.kt]: i think you should handle if the list is empty and throw exception

File: src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+        // then
+        assertFails {
+            useCase.addAudit(
+                audit.id, audit.createdBy, audit.entityId, audit.entityType,
+                audit.oldState, audit.newState, audit.date
+            )
+        }
+    }
+}
```
Comment (nourelhodaahmed) [inline src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt]: i think you should add test case
should throw Exception when auditRepository.addAuditLog throw Exception 
but if the coverage is 100, don't add it

File: src/main/kotlin/logic/model/Audit.kt [hunk 6]
```diff
@@ -1,14 +1,14 @@
 package squad.abudhabi.logic.model
 
 import java.time.LocalDate
+import java.util.UUID
 
 data class Audit(
-    val id: String,
-    val userId: String,
+    val id: String = UUID.randomUUID().toString(),
```
Comment (ElsayedMagdy122) [inline src/main/kotlin/logic/model/Audit.kt]: remove .toString , make type UUID

**Review by Noor-Serry — APPROVED** [review]

File: src/test/kotlin/data/audit/datasource/CsvAuditDataSourceTest.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    @BeforeEach
+    fun setup() {
+        csvFileHelper = mockk(relaxed = true)
+        csvAuditParser = mockk(relaxed = true)
+        csvAuditDataSource = CsvAuditDataSource(csvFileHelper, FILE_NAME, csvAuditParser)
+    }
+
+    @Test
+    fun `addAudit should call appendFile once when new audit is added successfully`(){
+
```
Comment (ElsayedMagdy122) [inline src/test/kotlin/data/audit/datasource/CsvAuditDataSourceTest.kt:32]: make comments start capital Given,When,Then

File: src/test/kotlin/CreateAuditHelper.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import squad.abudhabi.logic.model.Audit
+import squad.abudhabi.logic.model.EntityType
+import java.time.LocalDate
+import java.util.UUID
+
+fun createAudit(
+    id: String = UUID.randomUUID().toString(),
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/CreateAuditHelper.kt]: as bareq said in stockholm interview, we will use any id as UUID not string from UUID
so use UUID.randomUUID() instead of UUID.randomUUID().toString()

File: src/main/kotlin/data/audit/datasource/AuditDataSource.kt [hunk 9]
```diff
@@ -1,4 +1,9 @@
 package squad.abudhabi.data.audit.datasource
 
+import squad.abudhabi.logic.model.Audit
+
 interface AuditDataSource {
+
+    fun addAuditLog(audit: Audit)
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/audit/datasource/AuditDataSource.kt]: small change here let's make it create not add, for consistency across the app

File: src/main/kotlin/data/audit/datasource/CsvAuditDataSource.kt [hunk 10]
```diff
@@ -1,2 +1,22 @@
...
+import squad.abudhabi.data.utils.filehelper.CsvFileHelper
+import squad.abudhabi.logic.model.Audit
+
+class CsvAuditDataSource(
+    private val csvFileHelper: CsvFileHelper,
+    private val csvFileName: String,
+    private val csvAuditParser: CsvAuditParser,
+) : AuditDataSource{
+
+    override fun addAuditLog(audit: Audit) {
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/audit/datasource/CsvAuditDataSource.kt]: same here for naming

File: src/main/kotlin/data/audit/repository/AuditRepositoryImpl.kt [hunk 11]
```diff
@@ -1,2 +1,18 @@
...
 
+import squad.abudhabi.data.audit.datasource.AuditDataSource
+import squad.abudhabi.logic.model.Audit
+import squad.abudhabi.logic.repository.AuditRepository
+
+class AuditRepositoryImpl(
+    private val dataSource: AuditDataSource
+) : AuditRepository {
+
+    override fun addAuditLog(auditLog: Audit) {
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/data/audit/repository/AuditRepositoryImpl.kt]: also here

File: src/main/kotlin/logic/audit/AddAuditUseCase.kt [hunk 12]
```diff
@@ -1,2 +1,26 @@
-package squad.abudhabi.logic.audit
+package logic.audit
 
+import squad.abudhabi.logic.exceptions.InvalidAudit
+import squad.abudhabi.logic.model.Audit
+import squad.abudhabi.logic.repository.AuditRepository
+
+class AddAuditUseCase(
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/logic/audit/AddAuditUseCase.kt]: also here :")

File: src/main/kotlin/logic/audit/GetAuditUseCase.kt [hunk 13]
```diff
@@ -1,2 +1,18 @@
...
+
+class GetAuditUseCase(
+    private val auditRepository: AuditRepository
+){
+
+    fun getAuditHistory(entityId: String): List<Audit> {
+
+        if (entityId.isEmpty()) throw WrongInputException()
+        if (auditRepository.getAuditByEntityId(entityId).isEmpty()) throw EmptyList()
+        return auditRepository.getAuditByEntityId(entityId)
```
Comment (eslam-mohamed-dev) [inline src/main/kotlin/logic/audit/GetAuditUseCase.kt]: using FP is better here to avoid calling same code twice,
return auditRepository.getAuditByEntityId(entityId).takeIf { it.isNotEmpty() } ?: throw EmptyList()

File: src/test/kotlin/data/parser/CsvAuditParserTest.kt [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            entityId = "42",
+            entityType = EntityType.PROJECT,
+            oldState = "old",
+            newState = "new",
+            date = expectedDate
+        )
+
+        assertEquals(expectedAudit, result)
+    }
+
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/data/parser/CsvAuditParserTest.kt:77]: need to test failure also when throwing exceptions

File: src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+import io.mockk.mockk
+import io.mockk.verify
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.CsvSource
+import squad.abudhabi.logic.repository.AuditRepository
+import kotlin.test.assertFails
+
+class AddAuditUseCaseTest {
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt]: Taimor did you sleep? 😂
i just remind you of using create instead of add

File: src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        val audit = createAudit(
+            id = "213",
+            entityId = "asdww98"
+        )
+
+        // when
+        useCase.addAudit(audit)
+
+        // then
+        verify(exactly = 1) {
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt]: you can also verify fileHelper.appendFile() is called successfully

Comment (MohamedEmad7299) [reply src/test/kotlin/logic/audit/AddAuditUseCaseTest.kt]: i think it's not required here since i test it in data source

File: src/main/kotlin/logic/model/Audit.kt (hunk 6 above)
Comment (MohamedEmad7299) [reply src/main/kotlin/logic/model/Audit.kt:7]: and how i could handel it in parsing ?!

**Review by devnoorm7md — APPROVED** [review]

**Review by nourelhodaahmed — APPROVED** [review]

**Review by devnoorm7md — APPROVED** [review]

File: src/main/kotlin/presentation/audit/GetAuditUI.kt
Comment (eslam-mohamed-dev) [inline src/main/kotlin/presentation/audit/GetAuditUI.kt:1]: you can add 2 ui classes instead of 1 ui for many entity types
View History of a Project
View History of a Task

File: src/test/kotlin/data/audit/datasource/CsvAuditDataSourceTest.kt [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    @BeforeEach
+    fun setup() {
+        csvFileHelper = mockk(relaxed = true)
+        csvAuditParser = mockk(relaxed = true)
+        csvAuditDataSource = CsvAuditDataSource(csvFileHelper, FILE_NAME, csvAuditParser)
+    }
+
+    @Test
+    fun `addAudit should call appendFile once when new audit is added successfully`(){
```
Comment (eslam-mohamed-dev) [inline src/test/kotlin/data/audit/datasource/CsvAuditDataSourceTest.kt:31]: you forgot changing 'addAudit ' to 'createAudit' 😂

**Review by ElsayedMagdy122 — APPROVED** [review]
